src/mpid/ch4/netmod/ucx/ucx_win.c

```diff
@@ -68,6 +72,13 @@ static int win_allgather(MPIR_Win * win, size_t length, uint32_t disp_unit, void
         MPIDI_UCX_CHK_STATUS(status);
     }
 
+  map_check:
+    mpi_errno = MPIR_Allreduce(MPI_IN_PLACE, &map_failed, 1, MPI_C_BOOL, MPI_LOR, comm_ptr, &err);
```

Comment:
@minsii this adds another "all" collective into the window setup phase. We could probably piggyback this onto the allgather below if we convert `rkey_size` into an MPI_Aint and use `-1` for failing to map memory. What do you think?

Reply:
I don't think we have to fall back to AM for all processes when one of the process fails to map (e.g., only one process uses GPU memory as the winbuf but all others use host memory).

In `MPIDI_UCX_is_reachable_target`, we already check `MPIDI_UCX_WIN_INFO(win, rank).rkey`. Can we use the same field for the GPU winbuf case? E.g., by setting `rkey_size` to `-1` as you proposed.

We could then use a local for-loop at the end of `win_allgather` to check if all processes'  `MPIDI_UCX_WIN_INFO(win, rank).rkey` is NULL, then we could set `MPIDI_WINATTR_NM_REACHABLE` to avoid pointer dereference to `rkey` in `MPIDI_UCX_is_reachable_target`.

If we do the above change, we assume that for a single window, AM will be used for targets who use GPU winbuf and native RMA will still be used for targets who use host winbuf. Will there be any trouble if we do so?

If the above behavior is correct, we might want to change OFI to do the same thing too.

Reply:
Yes, that sounds like a better implementation. Let me see if I can hack it up in the next day or so.

Reply:
I think I'm going to take your approach for `main`, but a simpler approach for `3.4.x` that just handles the unsuccessful mapping and still relies on checking the `rkey` for reachability. Since `3.4.x` is a little higher priority, I'm going to open another PR and table this one for now.

Reply:
@minsii one consideration is that `ch4:ucx` currently sets `rkey = NULL` in the case where a process contributes no memory to the window, since it skips the `ucp_mem_map` part. I think it would be a regression for this to cause a window to fallback to active messages. Somehow we should distinguish between contributing no memory vs. "unreachable" memory.

One solution could be to always attempt to map the memory, even in the zero byte case. It looks like ucx will return a dummy handle and key for a zero length region.
https://github.com/openucx/ucx/blob/3c29f7ea214746d6c4a3df111b3e09d3f35792af/src/ucp/core/ucp_rkey.c#L126-L132
```
    if (memh->length == 0) {
        /* dummy memh, return dummy key */
        *rkey_buffer_p = &ucp_mem_dummy_buffer;
        *size_p        = sizeof(ucp_mem_dummy_buffer);
        status         = UCS_OK;
        goto out;
    }
```

Reply:
Another thing we could do is modify the condition like this:
```
if (winattr & MPIDI_WINATTR_NM_REACHABLE || win->rkey != NULL))
```
This was the case where all targets are reachable and the winattr is set, there is no dereference. Only in the mixed case will the right side fo the expression be evaluated, incurring the dereference.

Reply:
> Another thing we could do is modify the condition like this:
> 
> ```
> if (winattr & MPIDI_WINATTR_NM_REACHABLE || win->rkey != NULL))
> ```
> 
> This was the case where all targets are reachable and the winattr is set, there is no dereference. Only in the mixed case will the right side fo the expression be evaluated, incurring the dereference.

Is this the check in `MPIDI_UCX_is_reachable_target`? I agree.

Reply:
This change was actually less disruptive than I thought when I went to implement. I've updated both PRs with the same commits, including the reachability check improvement for `ch4:ucx`.